File: evidence/req-dem-021/edge-pulse.py
"""The EDGE pulse c0(t), evaluated from 3GPP TS 45.004 subclause 3.5 and nothing else.

This exists so that the C# filter is checked against the SPECIFICATION rather than against
itself. It is written from the published definition, in a different language, with a different
numerical method (quadrature rather than a closed form), and it emits a table of coefficients the
test compares against. If the two agree to 1e-6 across the pulse, two independent readings of the
standard agree; if they do not, one of them is wrong and the disagreement says where.

    c0(t) = product over i = 0..3 of S(t + iT),   for 0 <= t <= 5T, and 0 otherwise

    S(t) = sin(pi * integral from 0 to t of g)              for 0    <= t <= 4T
         = sin(pi/2 - pi * integral from 0 to t-4T of g)    for 4T   <= t <= 8T
         = 0                                                otherwise

    g(t) = 1/(2T) * [ Q(2*pi*0.3*(t - 5T/2) / (T*sqrt(ln 2)))
                    - Q(2*pi*0.3*(t - 3T/2) / (T*sqrt(ln 2))) ]

    Q(t) = 1/sqrt(2*pi) * integral from t to infinity of exp(-tau^2/2) dtau

BT = 0.3 is the 0.3 in g. The pulse is causal over [0, 5T] in the standard's own time reference;
the filter this project applies is centred, so the table below is c0(t + 5T/2) over |t| <= 5T/2.
"""

import math

import numpy as np

T = 1.0
BT = 0.3


def q(x):
    """The upper-tail Gaussian, 1/sqrt(2pi) * integral_x^inf exp(-tau^2/2) dtau."""
    return 0.5 * math.erfc(x / math.sqrt(2.0))


def g(t):
    a = 2.0 * math.pi * BT / (T * math.sqrt(math.log(2.0)))
    return (q(a * (t - 2.5 * T)) - q(a * (t - 1.5 * T))) / (2.0 * T)


START = -6.0
"""Where the phase integral is started, in symbol periods.

The standard writes the lower limit as 0. Taken literally that makes the pulse asymmetric by
1.2e-4, because g(0) is 1.7e-4 rather than zero -- while the mathematics says c0 is exactly
symmetric about 5T/2 (g is symmetric about 2T, so S is symmetric about 4T, and the product of four
shifted copies reflects onto itself). Starting six symbols lower, where g has fallen to about 1e-30,
gives symmetry to 2.7e-14. The two readings differ by up to 6.1e-5 -- above the 1e-6 the acceptance
criterion is stated at -- so the choice is deliberate and both are printed below."""


def integral_g(upper, steps=4096, lower=None):
    """Integral of g from 0 to upper by composite Simpson, written out rather than imported.

    Simpson on a smooth integrand converges as h^4; at 4096 panels over at most 4T the residual is
    far below the 1e-6 the comparison is made at, and doubling the panel count changes nothing in
    the twelfth decimal -- checked."""
    lo = START if lower is None else lower

    if upper <= lo:
        return 0.0

    if steps % 2:
        steps += 1

    h = (upper - lo) / steps
    total = g(lo) + g(upper)

    for k in range(1, steps):
        total += g(lo + k * h) * (4.0 if k % 2 else 2.0)

    return total * h / 3.0


def s(t, lower=None):
    if 0.0 <= t <= 4.0 * T:
        return math.sin(math.pi * integral_g(t, lower=lower))
    if 4.0 * T < t <= 8.0 * T:
        return math.sin(math.pi / 2.0 - math.pi * integral_g(t - 4.0 * T, lower=lower))
    return 0.0


def c0(t, lower=None):
    """The pulse in the standard's own time reference: non-zero on [0, 5T]."""
    if t < 0.0 or t > 5.0 * T:
        return 0.0
    product = 1.0
    for i in range(4):
        product *= s(t + i * T, lower=lower)
    return product


def centred(tau, lower=None):
    """The same pulse centred on zero, which is how a filter is applied."""
    return c0(tau + 2.5 * T, lower=lower)


if __name__ == "__main__":
    # T/8 over the whole support: 41 points, fine enough that a test interpolates nothing and few
    # enough that the table can sit in the test itself, where a reader can see what is being
    # compared against rather than being told a file exists.
    steps = 8
    taus = np.arange(-2.5 * steps, 2.5 * steps + 1) / steps

    values = np.array([centred(t) for t in taus])
    literal = np.array([centred(t, lower=0.0) for t in taus])

    print("# EDGE pulse c0(t), 3GPP TS 45.004 subclause 3.5, BT = 0.3")
    print("# Evaluated from the published definition alone: composite Simpson on g, nothing from")
    print("# any library beyond erfc, and no OpenVSA code. Column 1 is t/T, column 2 is c0(t+5T/2).")
    print("#")
    print("# THE LOWER LIMIT OF THE PHASE INTEGRAL IS A READING OF THE STANDARD:")
    print("#   from %+.0fT (this table): peak %.12f, asymmetry %.3e"
          % (START, values.max(), np.max(np.abs(values - values[::-1]))))
    print("#   from 0, as written:       peak %.12f, asymmetry %.3e"
          % (literal.max(), np.max(np.abs(literal - literal[::-1]))))
    print("#   the two differ by up to %.3e, against a 1e-6 criterion"
          % np.max(np.abs(values - literal)))
    print("# g(0) = %.6e per T, which is why the limit matters at all" % g(0.0))
    print("# area %.12f" % np.trapezoid(values, taus))
    for t, v in zip(taus, values):
        print("%+.6f	%.12f" % (t, v))

File: src/OpenVSA.Demod/Signal/EdgePulse.cs
using System;

namespace OpenVSA.Demod.Signal
{
    /// <summary>
    /// The EDGE transmit pulse: the linearised-GMSK main pulse <em>c₀(t)</em> of
    /// 3GPP TS 45.004 subclause 3.5 (<c>REQ-DEM-021</c>).
    /// </summary>
    /// <remarks>
    /// <para>
    /// <strong>It is not a Gaussian, and the requirement says so in a box of its own.</strong> It is
    /// the principal component of the Laurent decomposition of GMSK — the pulse that, driven by
    /// Dirac impulses, reproduces most of what a GMSK modulator does, and the one EDGE actually
    /// transmits its 3π/8-rotated 8PSK symbols through. A Gaussian at a fitted BT looks near enough
    /// on a plot to be believed; measured against this it is nowhere near, and a test says by how
    /// much.
    /// </para>
    /// <para>
    /// <strong>The standard's definition, verbatim in structure:</strong>
    /// </para>
    /// <code>
    /// c₀(t) = ∏(i = 0..3) S(t + iT),                       0 ≤ t ≤ 5T,  else 0
    ///
    /// S(t)  = sin(π ∫₀ᵗ g(t′) dt′),                        0 ≤ t ≤ 4T
    ///       = sin(π/2 − π ∫₀^(t−4T) g(t′) dt′),           4T ≤ t ≤ 8T
    ///       = 0                                            else
    ///
    /// g(t)  = 1/2T [ Q(2π·0.3 (t − 5T/2) / (T √ln2)) − Q(2π·0.3 (t − 3T/2) / (T √ln2)) ]
    ///
    /// Q(t)  = 1/√2π ∫ₜ^∞ e^(−τ²/2) dτ
    /// </code>
    /// <para>
    /// <strong>No quadrature: the integral of g has a closed form.</strong> g is a difference of two
    /// Q functions, and ∫Q(x)dx = xQ(x) − φ(x) with φ the standard normal density. The divergent
    /// parts of the two cancel, leaving <see cref="Phase"/> exact to the last bit — which matters
    /// because the pulse is a product of four of these and any error in one is multiplied into all
    /// of them.
    /// </para>
    /// <para>
    /// <strong>🔴 The lower limit of that integral is a reading of the standard, not a detail.</strong>
    /// The text writes ∫₀ᵗ, and g is not quite zero at t = 0 — it is 1.7e-4 — so taking the limit
    /// literally makes the pulse asymmetric by 1.2e-4, while the mathematics says c₀ is exactly
    /// symmetric about 5T/2 (g is symmetric about 2T, so S is symmetric about 4T, and the product of
    /// the four reflects onto itself). This implementation integrates from far enough below zero
    /// that g has genuinely vanished, which gives symmetry to 2.7e-14. <strong>The two readings
    /// differ by up to 6.1e-5 — larger than the 1e-6 the acceptance criterion is stated at — so the
    /// choice had to be made deliberately rather than fallen into.</strong> It is recorded on the
    /// issue and measured in <c>evidence/req-dem-021/</c>.
    /// </para>
    /// </remarks>
    internal static class EdgePulse
    {
        /// <summary>The bandwidth–time product in the standard's own definition of g.</summary>
        public const double BandwidthTime = 0.3;

        /// <summary>How far the pulse reaches either side of its centre, in symbols.</summary>
        /// <remarks>
        /// Two and a half. The standard defines c₀ on <c>0 ≤ t ≤ 5T</c> and applies it delayed by
        /// 2T; centred on its own axis of symmetry that is ±5T/2, and a filter span shorter than
        /// that truncates a pulse the transmitter did not truncate.
        /// </remarks>
        public const double ReachSymbols = 2.5;

        /// <summary>
        /// Where the phase integral is started, in symbols before zero.
        /// </summary>
        /// <remarks>
        /// Six symbols below the standard's own lower limit, where g has fallen to about 1e-30 — far
        /// past anything a double can tell from zero, so moving it further changes nothing. See the
        /// note on the class about why this is not simply zero.
        /// </remarks>
        private const double IntegralStartSymbols = -6.0;

        /// <summary>The pulse, centred on its own axis of symmetry.</summary>
        /// <param name="symbols">How far from the centre, in symbol periods.</param>
        /// <returns>The value of c₀(t + 5T/2); zero beyond ±5T/2.</returns>
        public static double At(double symbols)
        {
            if (Math.Abs(symbols) > ReachSymbols)
            {
                return 0.0;
            }

            return Causal(symbols + ReachSymbols);
        }

        /// <summary>The pulse in the standard's own time reference, non-zero on 0 ≤ t ≤ 5T.</summary>
        /// <param name="t">Time, in symbol periods from the start of the pulse.</param>
        internal static double Causal(double t)
        {
            if (t < 0.0 || t > 5.0)
            {
                return 0.0;
            }

            double product = 1.0;

            for (int i = 0; i < 4; i++)
            {
                product *= S(t + i);
            }

            return product;
        }

        /// <summary>The standard's S(t), in symbol periods.</summary>
        internal static double S(double t)
        {
            if (t < 0.0 || t > 8.0)
            {
                return 0.0;
            }

            if (t <= 4.0)
            {
                return Math.Sin(Math.PI * Phase(t));
            }

            return Math.Sin((Math.PI / 2.0) - (Math.PI * Phase(t - 4.0)));
        }

        /// <summary>
        /// The integral of g up to <paramref name="t"/>, in closed form.
        /// </summary>
        /// <param name="t">The upper limit, in symbol periods.</param>
        /// <returns>The phase integral, which runs from 0 to a half.</returns>
        /// <remarks>
        /// <para>
        /// g is <c>(Q(a(t − 5/2)) − Q(a(t − 3/2))) / 2</c> in symbol periods, and
        /// <c>∫Q(x)dx = xQ(x) − φ(x)</c>. Written for the difference, the parts that diverge at −∞
        /// cancel and what is left is exact:
        /// </para>
        /// <code>
        /// ∫ g = ( [x₂Q(x₂) − φ(x₂)] − [x₁Q(x₁) − φ(x₁)] ) / (2a) + 1/2   evaluated at the limits
        /// </code>
        /// <para>
        /// The constant is what the lower limit contributes: as t falls, both Q approach one and the
        /// bracket approaches <c>x₂ − x₁ = −a</c>, so the whole approaches −1/2 and the added half
        /// takes it to zero. That the result reaches exactly 1/2 as t grows is what makes
        /// <c>S(4T) = sin(π/2) = 1</c>, which is the property the two halves of S join on.
        /// </para>
        /// </remarks>
        internal static double Phase(double t)
        {
            double a = 2.0 * Math.PI * BandwidthTime / Math.Sqrt(Math.Log(2.0));

            return Antiderivative(a * (t - 2.5), a * (t - 1.5)) -
                Antiderivative(a * (IntegralStartSymbols - 2.5), a * (IntegralStartSymbols - 1.5));
        }

        /// <summary>The antiderivative of the difference of the two Q terms, at one point.</summary>
        private static double Antiderivative(double upper, double lower)
        {
            double a = 2.0 * Math.PI * BandwidthTime / Math.Sqrt(Math.Log(2.0));

            double first = (upper * Q(upper)) - Density(upper);
            double second = (lower * Q(lower)) - Density(lower);

            return (first - second) / (2.0 * a);
        }

        /// <summary>The standard normal density.</summary>
        private static double Density(double x) =>
            Math.Exp(-0.5 * x * x) / Math.Sqrt(2.0 * Math.PI);

        /// <summary>The upper-tail Gaussian, <c>Q(x) = erfc(x/√2)/2</c>.</summary>
        internal static double Q(double x) => 0.5 * Erfc(x / Math.Sqrt(2.0));

        /// <summary>
        /// The complementary error function, to about a part in 1e-15.
        /// </summary>
        /// <param name="x">The argument.</param>
        /// <returns><c>erfc(x)</c>.</returns>
        /// <remarks>
        /// <para>
        /// .NET Framework has no <c>erfc</c>, so it is here — a series near the origin where it
        /// converges quickly, and a continued fraction in the tail where the series does not. Both
        /// are the textbook forms, and the join at 2 is where each is comfortably inside its own
        /// range rather than at the edge of it.
        /// </para>
        /// <para>
        /// Accuracy matters more here than it looks: the pulse is a product of four values of S, so
        /// an error in one appears four times over, and the test that compares this filter with an
        /// independent evaluation of the standard is stated at 1e-6.
        /// </para>
        /// </remarks>
        internal static double Erfc(double x)
        {
            if (x < 0.0)
            {
                return 2.0 - Erfc(-x);
            }

            if (x < 2.0)
            {
                // erf(x) = 2/√π · e^(−x²) · Σ 2ⁿx^(2n+1) / (1·3·5···(2n+1)).
                double term = x;
                double sum = x;

                for (int n = 1; n < 200; n++)
                {
                    term *= 2.0 * x * x / ((2 * n) + 1);
                    sum += term;

                    if (Math.Abs(term) < Math.Abs(sum) * 1e-17)
                    {
                        break;
                    }
                }

                return 1.0 - (2.0 / Math.Sqrt(Math.PI) * Math.Exp(-x * x) * sum);
            }

            // erfc(x) = e^(−x²)/(x√π) · 1/(1 + 1/(2x²)/(1 + 2/(2x²)/(1 + …))), evaluated from the
            // bottom up. Sixty levels is far more than enough by x = 2, where the fraction's terms
            // are already shrinking quickly.
            double fraction = 0.0;

            for (int level = 60; level >= 1; level--)
            {
                fraction = level / 2.0 / (x + fraction);
            }

            return Math.Exp(-x * x) / (Math.Sqrt(Math.PI) * (x + fraction));
        }
    }
}
